New Workspaces

duranb opened the pull request:
`___REQUIRES_AERIE_PR___="1711"`

To test:

1. Navigate to http://localhost:3000/dictionaries
2. Upload at least a command dictionary (/e2e-tests/data/command-dictionary.xml)
3. Navigate to http://localhost:3000/parcels/new
4. Create a new parcel with the dictionaries uploaded in step 2
5. Navigate to http://localhost:3000/workspaces
6. Choose the parcel created in step 4
7. Specify a workspace directory location (cannot contain "/" in the directory name)
8. Specify a workspace name (optional)
9. Click the "Create" button
10. Click the "New Workspace Item" button: 
<img width="158" height="78" alt="Screenshot 2025-07-22 at 7 36 51 PM" src="https://github.com/user-attachments/assets/8585be3a-df40-483d-9e6f-e4efaa6c59b6" />

12. Verify that adding a new folder is correctly reflected in the workspace tree
13. Verify that adding a new sequence is correctly reflected in the workspace tree
14. Verify that importing a file is correctly reflected in the workspace tree
15. Verify that right clicking on tree items displays a context menu
16. Verify that the context menu actions function as expected
17. Click the "Grid" button on the sidebar:
<img width="67" height="71" alt="Screenshot 2025-07-22 at 7 42 33 PM" src="https://github.com/user-attachments/assets/0ad5ed9e-1ec2-4bd9-9de1-e7007d1611af" />

18. Verify that a tabular form of the workspace tree is displayed
19. Double click to open folders

AaronPlave commented on src/components/workspace/WorkspaceSidebar.svelte:
```diff
@@ -0,0 +1,344 @@
+                    {/if}
+                  </Tooltip.Trigger>
+                  <Tooltip.Content>
+                    <div class="text-sm">Refresh (last {getTimeAgo(lastRefreshTime, new Date())})</div>
```
The copy here feels a little terse - could it be something more like "Last refreshed X seconds ago"?
<img width="189" height="91" alt="image" src="https://github.com/user-attachments/assets/54e17573-50bc-4f62-965f-0311ac14087e" />

duranb replied:
I'm okay with that! @lklyne came up with it so I'll give him a chance to give a final 👍🏻 on this.

AaronPlave commented on src/components/workspace/WorkspaceSidebar.svelte:
```diff
@@ -0,0 +1,344 @@
+        <div class="grid h-full grid-rows-[min-content_auto]">
+          <Sidebar.Header className="p-0">
+            <div class="flex items-center justify-between gap-0 border-b border-border bg-background p-[8px]">
+              <SectionTitle>Workspace Table View</SectionTitle>
```
What is the reasoning for having distinct tabs for the table and the tree? Could this just be a single tab with a tree/grid switcher? Alternatively could there just be a tree with the ability to sort by name/date/etc as well as a filter on the tree?

AaronPlave replied:
Current implementation also has some duplicated code for the refresh and menu which would be good to avoid in any case (probably already be on your radar)

AaronPlave commented on src/components/workspace/WorkspacesGrid.svelte:
Can we get rid of the border on the grid here and replace it with just a top unrounded border?

AaronPlave commented on src/components/workspace/WorkspaceGridView/WorkspaceGridView.svelte:
Navigating away from a file with edits triggers the warning modal about unsaved edits however the selection in the grid changes to the new file even if you click "keep editing", resulting in a mismatched state. The tree view does not select the new file until the user confirms that they want to discard changes.

AaronPlave commented on src/components/ui/Sidebar/SidebarMenu/SidebarMenuSubButton.svelte:
```diff
@@ -0,0 +1,28 @@
+  data-slot="sidebar-menu-sub-button"
+  data-sidebar="menu-sub-button"
+  data-active={isActive}
+  class={cn(
```
The selected state of the tree view is very faint, could we make it stronger? I guarantee it'll be invisible over screen share for most. Also may want to make it blue to match grid view. And remove the border radius.

AaronPlave commented on src/components/modals/NewWorkspaceSequenceModal.svelte:
Consider doing a name conflict check on the user's new sequence name before allowing user to confirm creation. May not catch certain race conditions but may be decent reward vs effort for this PR?

duranb replied:
I probably won't be able to add this in, but this definitely should be added in for the next effort.

AaronPlave replied:
Can we make a workspace improvements ticket and dump this and any other punted items in there?

AaronPlave commented on src/components/modals/NewWorkspaceSequenceModal.svelte:
```diff
@@ -0,0 +1,98 @@
+          </Sidebar.Menu>
+        </Sidebar.Content>
+      </Sidebar.Provider>
+      <fieldset>
```
Might want to just use a div here with flex, gap, and vertical padding so that the field doesn't get the extra horizontal padding from the fieldset?

AaronPlave replied:
Though one unintended consequence is that given the slightly oversized focus states on stellar components the focus ring gets cut off. Could add a smidge of margin here to accommodate this. In the future I think we should just change this on the stellar side.

AaronPlave commented on src/components/ui/Sidebar/SidebarMenu/SidebarMenuSubButton.svelte:
Would be nice if possible to have the selected/hover button background appear behind the vertical hierarchy lines
<img width="115" height="199" alt="image" src="https://github.com/user-attachments/assets/8d9300f3-3dc4-477b-a2a6-f9b7cb50308e" />

duranb replied:
The only way I could really do this right now is to limit the extent of the background so it doesn't extend over the lines
<img width="308" height="50" alt="Screenshot 2025-07-23 at 2 24 53 PM" src="https://github.com/user-attachments/assets/c39a4a67-9916-41ff-9f98-1c9b6481179d" />

AaronPlave replied:
What about using a stronger blue with an opacity?

AaronPlave commented on src/components/workspace/WorkspaceSidebar.svelte:
```diff
@@ -0,0 +1,344 @@
+      <Tabs.List class="flex h-auto w-full flex-col items-center justify-start gap-0">
+        <Tooltip.Root>
+          <Tooltip.Trigger asChild let:builder>
+            <Tabs.Trigger value="files" class="flex h-10 w-10 items-center justify-center rounded-none shadow-none">
```
Any reason to keep this gray hover background on the active tab since clicking on the active tab has no effect?
<img width="51" height="106" alt="image" src="https://github.com/user-attachments/assets/d716aff7-7b1f-4a53-a867-f659e24dffd6" />

AaronPlave commented on src/components/workspace/WorkspaceSidebar.svelte:
Below ~768px in width the sidebar disappears. Know the app isn't really designed for such small screen sizes but feels odd that it is entirely gone / collapsed?
<img width="752" height="994" alt="image" src="https://github.com/user-attachments/assets/8df53a50-3765-447d-a728-1ddd626430c1" />

AaronPlave commented on src/assets/export.svg:
Could we use a lucide-icon for this? Also curious why the icon has changed here

AaronPlave commented on src/components/ui/Sidebar/SidebarMenu/SidebarMenuButton.svelte:
Thoughts on removing the vertical gap between items in the workspace tree?

AaronPlave commented on src/components/modals/ImportWorkspaceFileModal.svelte:
```diff
@@ -0,0 +1,104 @@
+<svelte:window on:keydown={onKeydown} />
+
+<Modal {height} {width}>
+  <ModalHeader on:close>Upload File(s) To Workspace</ModalHeader>
```
We should use either "Import" or "Upload" files but not both. The menu items in workspace say "Import file".

AaronPlave replied:
Also maybe we could just globally say "Upload files"

AaronPlave commented on src/components/workspace/WorkspaceGridView/WorkspaceGridView.svelte:
I'm all for icons for I actually think in this case we may want to not use icons in the context menu. I think the icons make the list of options a little harder to read, especially given the compactness. Also after looking other apps like IDEs, finder, etc., it looks like most of them don't have icons for every context menu option. Thoughts?
<img width="211" height="271" alt="image" src="https://github.com/user-attachments/assets/5a7a90f3-ed0e-4eb2-88a0-b80a4ad7e06d" />

AaronPlave commented on src/components/modals/RenameWorkspaceItemModal.svelte:
Renaming a file/folder to the same name causes an error. Might be good to prevent users from submitting the request if the name has not changed?

AaronPlave commented on src/components/workspace/WorkspaceSidebar.svelte:
```diff
@@ -0,0 +1,344 @@
+        <div class="grid h-full grid-rows-[min-content_auto]">
+          <Sidebar.Header className="p-0">
+            <div class="flex items-center justify-between gap-0 border-b border-border bg-background p-[8px]">
+              <SectionTitle>Workspace Tree View</SectionTitle>
```
This is a one pixel nit _but_ can we align the section title here (49px) with the height of the sequence editor and selected command panel (48px) 
<img width="526" height="66" alt="image" src="https://github.com/user-attachments/assets/e4f3a351-500f-4a66-b4d0-a7250a4d30e8" />

AaronPlave commented on src/components/modals/MoveItemToWorkspaceModal.svelte:
Managed to get this error when messing with the modal, will try to repro but may want to guard against this in the code 
<img width="828" height="355" alt="image" src="https://github.com/user-attachments/assets/d3280edb-20b2-4098-bf80-a2616ac91d71" />

AaronPlave replied:
Which seems to then break interactions with the tree

AaronPlave replied:
Seems to happen when I type in a different name for target directory

AaronPlave commented on src/components/workspace/Workspaces.svelte:
Would be nice to expand the file tree to the file in the url on load

duranb replied:
It shooould be doing this already. Do you mind screenshotting your workspace tree and the selected file in the URL? I'm wondering if there's a weird string matching bug.

AaronPlave replied:
Looks fixed

AaronPlave commented on src/components/modals/MoveItemToWorkspaceModal.svelte:
Kind of an odd message when i haven't selected a workspace yet - really this is just disabled because no workspace has been selected yet?
<img width="643" height="624" alt="image" src="https://github.com/user-attachments/assets/560873b4-d516-44a8-a552-a665c4c593a4" />

AaronPlave commented on src/components/modals/MoveWorkspaceItemModal.svelte:
Seeing odd `conflict` errors when i start moving files around even though I know there isn't a name conflict

AaronPlave replied:
https://github.com/user-attachments/assets/0951f8eb-344c-4149-a3da-978fc7fc82f3

duranb committed:
New Workspaces

duranb committed:
prevent table from selecting row if user decides to not navigate away from pending changes

duranb committed:
cleanup

duranb committed:
fix move between workspace permissions

duranb committed:
fix permission error when non-existing workspace is typed in

duranb committed:
fix tests pt 1

duranb committed:
change parcel loading listener in test

duranb committed:
bump up global playwright timeout

AaronPlave commented on src/routes/workspaces/[workspaceId]/+page.svelte:
```diff
@@ -0,0 +1,551 @@
+
+  async function goToSequence(filePath: string | null) {
+    if (updatedSelectedFileContent !== initialSelectedFileContent && selectedFilePath !== null) {
+      const { confirm } = await showConfirmModal(
```
Can't seem to click outside of the modal or hit escape to close it.

AaronPlave commented on src/components/workspace/WorkspaceTreeView/WorkspaceTreeView.svelte:
```diff
@@ -0,0 +1,268 @@
+      </ContextMenu.Item>
+      <ContextMenu.Separator />
+      <ContextMenu.Group>
+        <ContextMenu.Item size="sm" on:click={onNewSequence} aria-label="New Sequence">
```
Would be nice if clicking new sequence at the top would default to the user's current folder like the menu below it does, especially since the down arrow menu below is a little more hidden.

<img width="573" height="200" alt="Image" src="https://github.com/user-attachments/assets/167cfe34-d35f-4919-bda0-9d545ae86c7a" />

duranb replied:
This would be pretty easy to implement, but would it not be confusing if the same button does different things depending on the view? I can see that it would make sense to default to the current folder in the grid view because navigation is a little more deliberate and explicit. The tree view's navigation seems less so? I worry that they might unknowingly create the file in an unintended place (although it's not the hardest thing to choose the right place/undo).

AaronPlave replied:
Ah that's true, probably makes sense to have this dropdown be consistent across tree and grid views.

duranb committed:
bump up global playwright timeout

AaronPlave commented:
Would be nice to have an empty state for the tree view
<img width="408" height="536" alt="image" src="https://github.com/user-attachments/assets/e661b24f-9b93-4b81-ae63-37a1b01c6bdb" />

AaronPlave commented:
Edit: Didn't realize that the workspace name/folder was shown only the move/copy modals. No action needed but consider using a different icon for the workspace folder?
Managed to get into a bad state where the `Foo` folder is not shown in the workspace (tree and grid views) but is shown in the new sequence modal. Started from a new workspace and added `Foo` folder.
<img width="981" height="803" alt="image" src="https://github.com/user-attachments/assets/bf874516-6335-4d8e-b875-94732dc35568" />

AaronPlave commented:
Done: Could we remove the small empty element that exists when a folder is open but empty?
<img width="421" height="194" alt="image" src="https://github.com/user-attachments/assets/605164e8-8379-47da-a199-7ea86354d1d5" />

AaronPlave committed:
Styling

dandelany reviewed (changes requested):
A few notes from testing with @mattdailis today:

Major:
- If I create a file with no file extension, I can't open/edit it in the UI, and there is a JS error calling `split()` in the console
- In the Tree View, when a file is nested within a folder, the "..." icon (to open context menu) is not visible on the right side like it is for root-level elements
- When you click "create sequence", there is no clue in the UI that you need to name it with a specific extension (.seq in this case) in order to make it a sequence (may need to discuss)

Minor:
- There is "move" but not "Copy" in the file context menu, is this intentional/left for later?
- What determines ordering of files in the tree? It seems a bit random and inconsistent currently
- "workspace location" is confusing (during workspace creation), should this be required or not? any better label we could use, like "desired workspace folder name" or something? it's not clear to an uninitiated user what this controls.
- "Copy Link to File" is slightly confusing, could be "Copy Download Link"?

AaronPlave committed:
Refactor workspacegridview to use dropdowns instead of custom popovers

AaronPlave commented on src/components/ui/DataGrid/SingleActionDataGrid.svelte:
```diff
@@ -179,8 +201,8 @@
-        <ContextMenu.Item size="sm" disabled={!deletePermission} on:click={deleteItem}>
-          Delete {itemDisplayText}
+        <ContextMenu.Item class="items-center gap-1" size="sm" disabled={!deletePermission} on:click={deleteItem}>
+          <Trash2 size={16} /> Delete {itemDisplayText}{selectedItemId}
```
Any reason to display the selected item id here?

AaronPlave replied:
Or was it for debugging perhaps?

duranb replied:
Doh! Definitely leftover from debugging...

AaronPlave committed:
Style tweaks and fixes

duranb committed:
Make workspace modals dismissable
Rename "import file" to "upload file"
Rename "new sequence" to "new file"
add copy file/folder within same workspace
load unknown file types in TextEditor

duranb committed:
rename "location" to "folder name"
reword copy text

duranb commented:
@dandelany 

> Major:
> 
> * If I create a file with no file extension, I can't open/edit it in the UI, and there is a JS error calling `split()` in the console

Fixed!

> * In the Tree View, when a file is nested within a folder, the "..." icon (to open context menu) is not visible on the right side like it is for root-level elements

@AaronPlave fixed this!

> * When you click "create sequence", there is no clue in the UI that you need to name it with a specific extension (.seq in this case) in order to make it a sequence (may need to discuss)

Added the placeholder: "Enter file name (e.g. my_sequence.seq)"

> Minor:
> 
> * There is "move" but not "Copy" in the file context menu, is this intentional/left for later?

Added copy!

> * What determines ordering of files in the tree? It seems a bit random and inconsistent currently

Currently, the UI just displays the structure/ordering that the workspace service returns

> * "workspace location" is confusing (during workspace creation), should this be required or not? any better label we could use, like "desired workspace folder name" or something? it's not clear to an uninitiated user what this controls.

Workspace location is required by backend. I renamed it to "Workspace Folder Name"

> * "Copy Link to File" is slightly confusing, could be "Copy Download Link"?

If a file is clicked on, it will now say "Copy Download Link to File"; however if a folder is clicked, it will still say "Copy Link to Folder" since it's just to view the contents of the folder.

duranb committed:
fix incorrect selectors

duranb committed:
add delay before `setInputFiles` invocations

duranb committed:
fix dictionary test

duranb committed:
fix tests

duranb committed:
remove timeout

duranb committed:
add "copy" verbage to move modal

duranb committed:
extend timeouts for longer running test setups

duranb committed:
update test docker compose...

duranb committed:
add volume

dandelany committed:
fix lint issues in docker-compose

duranb committed:
update testing docs for new services

duranb committed:
fix workspace creation state reset

duranb committed:
fix channel dictionary setting logic

dandelany committed:
update aerie-actions library to v1.0.0

dandelany reviewed (approved):
All issues above have been addressed and tests are passing :tada: I've also tested it thoroughly alongside https://github.com/NASA-AMMOS/aerie/pull/1711 locally.

This is ready to merge. Thanks @duranb @AaronPlave !

duranb commented:
I believe this also resolves #1483
